File: benchmark/06_raytracer/raytracer.go
// Benchmark 06: Simple Ray Tracer
// Category: Simulation / Floating Point
// Tests: 3D math, vector operations, recursion

package main

import (
	"fmt"
	"math"
	"time"
)

const (
	WIDTH     = 400
	HEIGHT    = 400
	MAX_DEPTH = 4
)

type Vec3 struct {
	X, Y, Z float64
}

type Sphere struct {
	Center Vec3
	Radius float64
}

func vec3Add(a, b Vec3) Vec3 {
	return Vec3{a.X + b.X, a.Y + b.Y, a.Z + b.Z}
}

func vec3Sub(a, b Vec3) Vec3 {
	return Vec3{a.X - b.X, a.Y - b.Y, a.Z - b.Z}
}

func vec3Mul(a Vec3, t float64) Vec3 {
	return Vec3{a.X * t, a.Y * t, a.Z * t}
}

func vec3Dot(a, b Vec3) float64 {
	return a.X*b.X + a.Y*b.Y + a.Z*b.Z
}

func vec3Length(a Vec3) float64 {
	return math.Sqrt(vec3Dot(a, a))
}

func vec3Normalize(a Vec3) Vec3 {
	len := vec3Length(a)
	return vec3Mul(a, 1.0/len)
}

// Ray-sphere intersection
func hitSphere(s Sphere, origin, dir Vec3) float64 {
	oc := vec3Sub(origin, s.Center)
	a := vec3Dot(dir, dir)
	b := 2.0 * vec3Dot(oc, dir)
	c := vec3Dot(oc, oc) - s.Radius*s.Radius
	discriminant := b*b - 4*a*c

	if discriminant < 0 {
		return -1.0
	}
	return (-b - math.Sqrt(discriminant)) / (2.0 * a)
}

// Trace ray and return color intensity
func trace(origin, dir Vec3, spheres []Sphere, depth int) float64 {
	if depth <= 0 {
		return 0.0
	}

	closest := 1e10
	hitIdx := -1

	for i, s := range spheres {
		t := hitSphere(s, origin, dir)
		if t > 0.001 && t < closest {
			closest = t
			hitIdx = i
		}
	}

	if hitIdx < 0 {
		return 0.0
	}

	hitPoint := vec3Add(origin, vec3Mul(dir, closest))
	normal := vec3Normalize(vec3Sub(hitPoint, spheres[hitIdx].Center))
	reflected := vec3Sub(dir, vec3Mul(normal, 2.0*vec3Dot(dir, normal)))

	return 0.5 + 0.5*trace(hitPoint, reflected, spheres, depth-1)
}

func main() {
	start := time.Now()

	// Scene setup
	spheres := []Sphere{
		{Center: Vec3{0, 0, -5}, Radius: 1.0},
		{Center: Vec3{2, 0.5, -4}, Radius: 0.5},
		{Center: Vec3{-2, 0, -6}, Radius: 0.7},
	}

	totalIntensity := 0.0

	// Render
	for y := 0; y < HEIGHT; y++ {
		for x := 0; x < WIDTH; x++ {
			u := (2.0*float64(x) - float64(WIDTH)) / float64(HEIGHT)
			v := (float64(HEIGHT) - 2.0*float64(y)) / float64(HEIGHT)
			dir := vec3Normalize(Vec3{u, v, -1.0})
			origin := Vec3{0, 0, 0}

			intensity := trace(origin, dir, spheres, MAX_DEPTH)
			totalIntensity += intensity
		}
	}

	elapsed := time.Since(start)

	fmt.Printf("Image size: %dx%d\n", WIDTH, HEIGHT)
	fmt.Printf("Average intensity: %.6f\n", totalIntensity/float64(WIDTH*HEIGHT))
	fmt.Printf("Time: %.4f seconds\n", elapsed.Seconds())
}

File: benchmark/06_raytracer/raytracer.rs
// Benchmark 06: Simple Ray Tracer
// Category: Simulation / Floating Point
// Tests: 3D math, vector operations, recursion

use std::time::Instant;

const WIDTH: usize = 400;
const HEIGHT: usize = 400;
const MAX_DEPTH: usize = 4;

#[derive(Clone, Copy)]
struct Vec3 {
    x: f64,
    y: f64,
    z: f64,
}

struct Sphere {
    center: Vec3,
    radius: f64,
}

impl Vec3 {
    fn new(x: f64, y: f64, z: f64) -> Self {
        Vec3 { x, y, z }
    }

    fn add(self, other: Vec3) -> Vec3 {
        Vec3::new(self.x + other.x, self.y + other.y, self.z + other.z)
    }

    fn sub(self, other: Vec3) -> Vec3 {
        Vec3::new(self.x - other.x, self.y - other.y, self.z - other.z)
    }

    fn mul(self, t: f64) -> Vec3 {
        Vec3::new(self.x * t, self.y * t, self.z * t)
    }

    fn dot(self, other: Vec3) -> f64 {
        self.x * other.x + self.y * other.y + self.z * other.z
    }

    fn length(self) -> f64 {
        (self.dot(self)).sqrt()
    }

    fn normalize(self) -> Vec3 {
        self.mul(1.0 / self.length())
    }
}

// Ray-sphere intersection
fn hit_sphere(s: &Sphere, origin: Vec3, dir: Vec3) -> f64 {
    let oc = origin.sub(s.center);
    let a = dir.dot(dir);
    let b = 2.0 * oc.dot(dir);
    let c = oc.dot(oc) - s.radius * s.radius;
    let discriminant = b * b - 4.0 * a * c;

    if discriminant < 0.0 {
        return -1.0;
    }
    (-b - discriminant.sqrt()) / (2.0 * a)
}

// Trace ray and return color intensity
fn trace(origin: Vec3, dir: Vec3, spheres: &[Sphere], depth: usize) -> f64 {
    if depth == 0 {
        return 0.0;
    }

    let mut closest = 1e10;
    let mut hit_idx = None;

    for (i, s) in spheres.iter().enumerate() {
        let t = hit_sphere(s, origin, dir);
        if t > 0.001 && t < closest {
            closest = t;
            hit_idx = Some(i);
        }
    }

    if let Some(idx) = hit_idx {
        let hit_point = origin.add(dir.mul(closest));
        let normal = hit_point.sub(spheres[idx].center).normalize();
        let reflected = dir.sub(normal.mul(2.0 * dir.dot(normal)));

        0.5 + 0.5 * trace(hit_point, reflected, spheres, depth - 1)
    } else {
        0.0
    }
}

fn main() {
    let start = Instant::now();

    // Scene setup
    let spheres = [
        Sphere {
            center: Vec3::new(0.0, 0.0, -5.0),
            radius: 1.0,
        },
        Sphere {
            center: Vec3::new(2.0, 0.5, -4.0),
            radius: 0.5,
        },
        Sphere {
            center: Vec3::new(-2.0, 0.0, -6.0),
            radius: 0.7,
        },
    ];

    let mut total_intensity = 0.0;

    // Render
    for y in 0..HEIGHT {
        for x in 0..WIDTH {
            let u = (2.0 * x as f64 - WIDTH as f64) / HEIGHT as f64;
            let v = (HEIGHT as f64 - 2.0 * y as f64) / HEIGHT as f64;
            let dir = Vec3::new(u, v, -1.0).normalize();
            let origin = Vec3::new(0.0, 0.0, 0.0);

            let intensity = trace(origin, dir, &spheres, MAX_DEPTH);
            total_intensity += intensity;
        }
    }

    let elapsed = start.elapsed();

    println!("Image size: {}x{}", WIDTH, HEIGHT);
    println!("Average intensity: {:.6}", total_intensity / (WIDTH * HEIGHT) as f64);
    println!("Time: {:.4} seconds", elapsed.as_secs_f64());
}
